feat: multi-prev (#145)
ceramicnetwork/CIPs: CIPs/cip-145.md

---
cip: 145
title: Multiple previous (multi-prev)
author: Aaron Goldman (@aarongoldman), Mohsin Zaidi (@smrz2001)
discussions-to: https://forum.ceramic.network/t/cip-145-multiple-previous-multi-prev/1277
status: Draft

> category: <Core>

Reviewer: ```suggestion
category: Core
```

---
cip: 145
title: Multiple previous (multi-prev)
author: Aaron Goldman (@aarongoldman), Mohsin Zaidi (@smrz2001)
discussions-to: <URL to a topic under the CIP category on the Ceramic forum: https://forum.ceramic.network/c/cips>
status: Draft
category: <Core | Networking | Interface | RFC | Meta>
created: 2023-08-21
edited: 2023-08-21

> requires: <EIP number(s)>
> replaces: <EIP number(s)>

Reviewer: ```suggestion
```

<!-- STEPS TO SUBMIT A CIP:
1. Complete the header above.
2. Fill in as much content as is appropriate for the status of your CIP.-->

<!--ADDITIONAL INSTRUCTIONS FOR HEADER SECTION ABOVE-->

<!--[title]: Give your issue a concise, descriptive title prefixed by either its *type* for standards CIPs or its category for other CIPs. (i.e. Core: Protocol Upgrade, Meta: Define CIP Process, etc.).-->

<!--[category]: Here is a description of category terms.
- `Core`: an CIP that affects the core protocol.
- `Networking`: an CIP thst affects the networking layer (i.e. libp2p or syncing).
- `Interface`: an CIP that affects the Ceramic API or provider interface.
- `RFC`: an CIP that proposes an implementation standard (i.e. doctypes, document configurations, or document schemas).
- `Meta`: an CIP that affects the governance process for CIPs.-->

<!--[requires]: A list of CIP(s) that this CIP depends on. *Optional.-->

<!--[replaces]: A list of CIP(s) that this CIP replaces. *Optional.-->

## Simple Summary
<!--Provide a simplified and layman-accessible explanation of the CIP.-->

> Multi-prev allows the `priv` field in a ceramic event header to be a list of string rather then a string. This is an ordered list and the abrogator may treat the same set of priv CIDs differently depending on the order.

Reviewer: ```suggestion
Multi-prev allows the `prev` field in a ceramic event header to be a list of CIDs rather then a single CID. This is an ordered list and the abrogator may treat the same set of prev CIDs differently depending on the order.
```

These are stored as CID type not string.

* A join `A ⋁ B` find the common ancestor of the events `A` and `B`
* A meet `A ⋀ B` find the common descendent of the events `A` and `B`

## Specification
<!--The technical specification should describe the syntax and semantics of any new feature.-->
Specification goes here.

## Rationale
<!--The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion.-->
Rationale goes here.

## Backwards Compatibility
<!--All CIPs that introduce backwards incompatibilities must include a section describing these incompatibilities and their severity. The CIP must explain how the author proposes to deal with these incompatibilities. CIP submissions without a sufficient backwards compatibility section may be rejected outright.-->
All events that have a string in `prev` instead of a list will work with this CIP.

> Nodes that don't have this CIP will not understand what new events with a `prev` list means.

Reviewer: So it's a soft fork? New nodes will be able to understand old nodes, but not the other way around.

Reviewer 2: Yes, new nodes should work with everything that already exists.

Please note that new data even on existing streams might be aggregated differently on newer nodes compared to older nodes.

## Implementation
<!--The implementations must be completed before any CIP is given status "Final", but it need not be completed before the CIP is accepted.-->
Implementation goes here.

> legacy `priv`

Reviewer: ```suggestion
legacy `prev`
```

```json
{ "prev" = [CID, CID, CID]}
```

> multi-priv `priv`

Reviewer: ```suggestion
multi-priv `prev`
```
